src/loader/loader.c

```diff
@@ -91,41 +91,71 @@ drop_statement_drops_extension(DropStmt *stmt)
 }
 
 static bool
-load_utility_cmd(Node *utilityStmt)
+handle_variable_set(Node *utility_stmt)
 {
-	switch (nodeTag(utilityStmt))
-	{
-		case T_VariableSetStmt:
-			if (strcmp(((VariableSetStmt *) utilityStmt)->name, GUC_DISABLE_LOAD_NAME) == 0)
-				/* Do not load when setting the guc */
-				return false;
-			return true;
-		case T_AlterExtensionStmt:
-			if (strcmp(((AlterExtensionStmt *) utilityStmt)->extname, EXTENSION_NAME) != 0)
-				return true;
+	VariableSetStmt *stmt = (VariableSetStmt *) utility_stmt;
 
-			/* disallow loading two .so from different versions */
-			if (loaded)
-				ereport(ERROR,
-						(errmsg("Cannot update the extension after the old version has already been loaded"),
-						 errhint("You should start a new session and execute ALTER EXTENSION as the first command")));
+	/* Do not load when setting the guc */
+	return strcmp(stmt->name, GUC_DISABLE_LOAD_NAME) != 0;
+}
 
-			/* do not load the current (old) version's .so */
-			return false;
-		case T_CreateExtensionStmt:
-			if (!loaded || strcmp(((CreateExtensionStmt *) utilityStmt)->extname, EXTENSION_NAME) != 0)
-				return true;
+static bool
+handle_alter_extension(Node *utility_stmt)
+{
+	AlterExtensionStmt *stmt = (AlterExtensionStmt *) utility_stmt;
 
-			/* disallow loading two .so from different versions */
-			ereport(ERROR,
-					(errmsg("Cannot create the extension after the another version has already been loaded"),
-					 errhint("You should start a new session and execute CREATE EXTENSION as the first command")));
+	if (strcmp(stmt->extname, EXTENSION_NAME) != 0)
+		return true;
+
+	/* disallow loading two .so from different versions */
+	if (loaded)
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("extension \"%s\" cannot be updated after the old version has already been loaded", stmt->extname),
+				 errhint("You should start a new session and execute ALTER EXTENSION as the first command")));
+
+	/* do not load the current (old) version's .so */
+	return false;
+}
 
+static bool
+handle_create_extension(Node *utility_stmt)
+{
+	CreateExtensionStmt *stmt = (CreateExtensionStmt *) utility_stmt;
+	bool		is_extension = strcmp(stmt->extname, EXTENSION_NAME) == 0;
+
+	if (!loaded || !is_extension)
+		return true;
+
+	if (stmt->if_not_exists)
```

Comment:
What happens if you have an extension loaded, drop the extension, and create it again with a different version using `IF NOT EXISTS`. My reading of this code is that it will allow you to do so even though that is wrong. I think we should have this case tested.